Add Watchdog HAL API specification

scartmell-arm opened the pull request:
# Watchdog API

HAL watchdog functionality will be implemented as two separate APIs. The reset reason API allows a user to detect the last system reset reason to determine if a Watchdog was triggered. The Watchdog API allows configuring and updating Watchdog timers on all boards.

## Reset Reason API
mbedOS requires a standardized API for reading and resetting the reset reason for an mbedOS development board in order to usefully implement the Watchdog API.

When a watchdog reset is triggered it is detectable by reading the reset reason register at boot time to make corrective actions. There is currently no standardized method of reading the reset reason for a board in mbedOS. Each vendor implements its own set of functions for reading and resetting the reset reason for board families.  

The reset reason API should provide access to a common set of functionality provided by some of the common mbedOS boards.

## Watchdog API
mbedOS requires an API for configuring, running, and operating the hardware watchdogs across a number of development boards.

Watchdogs can be used to detect erroneous operation of the application, and at boot time utilize the reset reason API to determine if a watchdog caused the reset and take corrective actions.

There is also, currently, no standard method of initializing, configuring, and running a watchdog timer.

Commit:
Add Watchdog HAL API specification headers

HAL watchdog functionality will be implemented as two separate APIs. The reset
reason API allows a user to detect the last system reset reason to identify if
a Watchdog was triggered. The Watchdog API allows configuring and updating
Watchdog timers on all boards. This commit defines the headers.

bulislaw reviewed (changes requested):
Also I think the PR should be against master we add new optional APIs they won't break nobody's tests.

Commit:
Add K64F reset reason reference implementation

Commit:
Add simple watchdog reference implementation for K64F

- Add preprocessor guard to watchdog api that errors if the reset reason api is not also implemented
- Add RESET_REASON and WATCHDOG to K64F targets.json
- Add watchdog reference implementation

Commit:
Fix watchdog API issues

- Fix typo in module comment
- Redefine the default system behaviour in sleep mode
- Guard K64F enableWait flag
- Remove bit shifts from reset reason enum

Commit:
Add Reset Reason platform API

Commit:
Add reset reason reference implementation STM32

Commit:
Add function to fetch platform specific reset reason register values

bulislaw reviewed (commented):
I think `platform/mbed_reset_reason.cpp` should be moved to `drivers/` cause it's a driver in our current terminology. Otherwise looks good.

Commit:
Add independent watchdog reference implementation for STM32

bulislaw reviewed (commented):
For consistency reason we should implement both reset reason and watchdog drivers as classes, we can use static member function if that makes more sense.

Commit:
Amend reset reason driver API
- Change API to match C++ API throughout drivers
- Amend HAL API documentation to be more specific

Commit:
Add Watchdog driver API

sg- reviewed (dismissed):
Did we do a study of the different device watchdog capabilities? From reading the code it doesnt look like it but if so please point me at where this can be found.

Commit:
Remove window and sleep mode options for watchdog API

Commit:
Add missing license headers

Commit:
Move watchdog parameter validation into the driver layer

0xc0170 reviewed (approved):
Once all files have header files, and possibly handling input parameters in HAL is consolidated (common checks can be handled in the upper layer)

Commit:
Fix Watchdog::stop on K64F target

WDOG_Disable will suspend the watchdog until a reset is triggered instead of stopping it altogether. Deinit will disable it until it is reinitialized.

fkjagodzinski committed:
K64F: watchdog HAL: Fix max_timeout.

Limit MAX_TIMEOUT_MS to UINT32_MAX.
Improve MAX_TIMEOUT_MS handling.
Fix hal_watchdog_get_platform_features() returning
.max_timeout = 0xfffffff8 instead of 0xffffffff.

fkjagodzinski reviewed (commented):
Found a few minor issues. If you find this useful, here is my patch c0d0ca8b898a98b340f35d754db2f12e8db8842f that you can cherry-pick.

Commit:
Add RESET_REASON and WATCHDOG definitions to hexiwear build target

Commit:
Rename watchdog.c -> watchdog_api.c to prevent name collision with Platform API

SenRamakri reviewed (approved):
Looks like most of the issues I thought of commenting has been captured/addressed in some way already. I'll approve this merging into feature branch.
